Build failure with GHC 9.2.1

```

src/Apecs/THTuples.hs:94:30: error:
    • Couldn't match type ‘Pat’ with ‘Type’
      Expected: Name -> Type
        Actual: Name -> Pat
    • In the first argument of ‘(<$>)’, namely ‘VarP’
      In the second argument of ‘ConP’, namely ‘(VarP <$> sNs)’
      In the expression: ConP tupleName (VarP <$> sNs)
   |
94 |       sPat = ConP tupleName (VarP <$> sNs)
   |                              ^^^^

src/Apecs/THTuples.hs:100:30: error:
    • Couldn't match type ‘Pat’ with ‘Type’
      Expected: Name -> Type
        Actual: Name -> Pat
    • In the first argument of ‘(<$>)’, namely ‘VarP’
      In the second argument of ‘ConP’, namely ‘(VarP <$> wNs)’
      In the expression: ConP tupleName (VarP <$> wNs)
    |
100 |       wPat = ConP tupleName (VarP <$> wNs)
    |                              ^^^^

src/Apecs/THTuples.hs:133:30: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘ConP’ is applied to too few arguments
      In the first argument of ‘Match’, namely ‘(ConP 'False [])’
      In the expression:
        Match
          (ConP 'False []) (NormalB $ AppE (VarE 'return) (ConE 'False)) []
      In the first argument of ‘LamCaseE’, namely
        ‘[Match
            (ConP 'False []) (NormalB $ AppE (VarE 'return) (ConE 'False)) [],
          Match (ConP 'True []) (NormalB vb) []]’
    |
133 |           (LamCaseE [ Match (ConP 'False []) (NormalB$ AppE (VarE 'return) (ConE 'False)) []
    |                              ^^^^^^^^^^^^^^

src/Apecs/THTuples.hs:134:30: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘ConP’ is applied to too few arguments
      In the first argument of ‘Match’, namely ‘(ConP 'True [])’
      In the expression: Match (ConP 'True []) (NormalB vb) []
      In the first argument of ‘LamCaseE’, namely
        ‘[Match
            (ConP 'False []) (NormalB $ AppE (VarE 'return) (ConE 'False)) [],
          Match (ConP 'True []) (NormalB vb) []]’
    |
134 |                     , Match (ConP 'True []) (NormalB vb) []
    |                              ^^^^^^^^^^^^^

src/Apecs/THTuples.hs:140:34: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘sPat’ is applied to too few arguments
      In the expression: sPat
      In the first argument of ‘Clause’, namely ‘[sPat, etyPat]’
      In the expression:
        Clause
          [sPat, etyPat] (NormalB $ liftAll tuplE (explGetF <$> sEs)) []
    |
140 |         [ FunD explGetN [Clause [sPat, etyPat]
    |                                  ^^^^

src/Apecs/THTuples.hs:144:37: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘sPat’ is applied to too few arguments
      In the expression: sPat
      In the first argument of ‘Clause’, namely ‘[sPat, etyPat]’
      In the expression:
        Clause
          [sPat, etyPat]
          (NormalB
             $ foldr
                 explExistsAnd (AppE (VarE 'pure) (ConE 'True))
                 ((`AppE` etyE) . explExistsF <$> sEs))
          []
    |
144 |         , FunD explExistsN [Clause [sPat, etyPat]
    |                                     ^^^^

src/Apecs/THTuples.hs:150:34: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘sPat’ is applied to too few arguments
      In the expression: sPat
      In the first argument of ‘Clause’, namely ‘[sPat, etyPat, wPat]’
      In the expression:
        Clause
          [sPat, etyPat, wPat]
          (NormalB $ sequenceAll (zipWith explSetF sEs wEs)) []
    |
150 |         [ FunD explSetN [Clause [sPat, etyPat, wPat]
    |                                  ^^^^

src/Apecs/THTuples.hs:150:48: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘wPat’ is applied to too few arguments
      In the expression: wPat
      In the first argument of ‘Clause’, namely ‘[sPat, etyPat, wPat]’
      In the expression:
        Clause
          [sPat, etyPat, wPat]
          (NormalB $ sequenceAll (zipWith explSetF sEs wEs)) []
    |
150 |         [ FunD explSetN [Clause [sPat, etyPat, wPat]
    |                                                ^^^^

src/Apecs/THTuples.hs:156:38: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘sPat’ is applied to too few arguments
      In the expression: sPat
      In the first argument of ‘Clause’, namely ‘[sPat, etyPat]’
      In the expression:
        Clause
          [sPat, etyPat] (NormalB $ sequenceAll (explDestroyF <$> sEs)) []
    |
156 |         [ FunD explDestroyN [Clause [sPat, etyPat]
    |                                      ^^^^

src/Apecs/THTuples.hs:162:38: error:
    • Couldn't match expected type ‘Pat’
                  with actual type ‘[Pat] -> Pat’
    • Probable cause: ‘sPat’ is applied to too few arguments
      In the expression: sPat
      In the first argument of ‘Clause’, namely ‘[sPat]’
      In the expression:
        Clause
          [sPat]
          (NormalB
             $ foldl
                 explMembersFold (explMembersF (head sEs))
                 (explExistsF <$> tail sEs))
          []
    |
162 |         [ FunD explMembersN [Clause [sPat]
    |                                      ^^^^
cabal: Failed to build apecs-0.9.3.
```

I got the same errors with v0.8.3 and up.

As a Hackage trustee, I have created some Hackage revisions that should prevent users from getting these errors. See e.g. https://hackage.haskell.org/package/apecs-0.8.3/revisions/.

--- a/.github/workflows/CI.yaml
+++ b/.github/workflows/CI.yaml
@@ -24,18 +24,17 @@ jobs:
         run: nix build -L -f shell.nix
 
   stack-build:
-    name: Stack ${{ matrix.resolver }} on ${{ matrix.os }} (${{ matrix.required }})
+    name: Stack ${{ matrix.resolver }} on ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
-        resolver: [ lts-16, lts-15, lts-14 ]
-        os: [ macos-latest, ubuntu-latest ]
-        required: [ required ]
+        os: [ ubuntu-latest, macos-10.15 ]
+        resolver: [ lts-18, lts-17, lts-16, lts-15, lts-14 ]
         include:
-          - resolver: nightly
-            os: ubuntu-latest
-            required: optional
-    continue-on-error: ${{ matrix.required == 'optional' }}
+          - os: macos-latest
+            resolver: lts-17
+          - os: macos-latest
+            resolver: lts-18
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -49,25 +48,22 @@ jobs:
             ~/.stack
             .stack-work
           key: stack-${{ matrix.os }}-${{ matrix.resolver }}-${{ hashFiles('**/*.cabal', 'stack.yaml') }}
-      - run: sudo apt-get install freeglut3-dev
+      - run: sudo apt-get update && sudo apt-get install freeglut3-dev
         if: matrix.os == 'ubuntu-latest'
       - run: stack build --only-dependencies --resolver ${{ matrix.resolver }}
       - run: stack build --resolver ${{ matrix.resolver }}
       - run: stack test --resolver ${{ matrix.resolver }}
 
   cabal-build:
-    name: Cabal GHC ${{ matrix.ghc }} on ${{ matrix.os }} (${{ matrix.required }})
+    name: Cabal GHC ${{ matrix.ghc }} on ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
-        ghc: [ '8.6.5', '8.8.3', '8.10.2' ]
-        os: [ ubuntu-latest, macos-latest ]
-        required: [ required ]
+        ghc: [ '8.6.5', '8.8.3', '8.10.7' ]
+        os: [ ubuntu-latest, macos-10.15 ]
         include:
-          - ghc: latest
-            os: ubuntu-latest
-            required: optional
-    continue-on-error: ${{ matrix.required == 'optional' }}
+          - os: macos-latest
+            ghc: '8.10.7'
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v2
@@ -80,8 +76,33 @@ jobs:
             ~/.cabal
             dist-newstyle
           key: cabal-cache-${{ matrix.os }}-${{ matrix.ghc }}-${{ hashFiles('**/*.cabal') }}
-      - run: sudo apt-get install freeglut3-dev
+      - run: sudo apt-get update && sudo apt-get install -y freeglut3-dev
         if: matrix.os == 'ubuntu-latest'
       - run: cabal new-build all --only-dependencies
       - run: cabal new-build all
       - run: cabal new-test all --test-show-details=direct
+
+  cabal-build-ghc9:
+    name: apecs, apecs-physics GHC ${{ matrix.ghc }} on ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        ghc: [ '9.0.1', '9.2.1' ]
+        os: [ ubuntu-latest, macos-10.15, macos-latest ]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: haskell/actions/setup@v1
+        with:
+          ghc-version: ${{ matrix.ghc }}
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cabal
+            dist-newstyle
+          key: cabal-cache-${{ matrix.os }}-${{ matrix.ghc }}-${{ hashFiles('**/*.cabal') }}
+      - run: sudo apt-get update && sudo apt-get install -y freeglut3-dev
+        if: matrix.os == 'ubuntu-latest'
+      - run: cabal new-build all --only-dependencies --project-file=cabal-ghc9.project
+      - run: cabal new-build all --project-file=cabal-ghc9.project
+      - run: cabal new-test all --test-show-details=direct --project-file=cabal-ghc9.project

--- a/.github/workflows/CI.yaml
+++ b/.github/workflows/CI.yaml
@@ -9,17 +9,17 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
-      - uses: cachix/install-nix-action@v12
-      - uses: cachix/cachix-action@v8
+      - uses: cachix/install-nix-action@v16
+      - uses: cachix/cachix-action@v10
         with:
           name: jmc
           authToken: '${{ secrets.CACHIX_AUTH_TOKEN }}'
       - name: Build dependencies
-        run: nix build -L '(import ./pkgs.nix).hsPkgs.shellFor {}'
+        run: nix build -L --impure --expr '(import ./pkgs.nix).hsPkgs.shellFor {}'
       - name: Build apecs
         run: nix build -L -f pkgs.nix hsPkgs.apecs.checks
       - name: Build all local packages
-        run: nix build -L
+        run: nix build -L -f default.nix
       - name: Build dev enviroment
         run: nix build -L -f shell.nix
 

--- a/apecs-physics/src/Apecs/Physics/Collision.hs
+++ b/apecs-physics/src/Apecs/Physics/Collision.hs
@@ -155,6 +155,6 @@ addPostStepCallback (toEnum -> k) systemCallback= do
     let fn = castFunPtrToPtr funPtr
     [C.block| void {
       int *data = 0;
-      cpSpaceAddPostStepCallback($(cpSpace *space), $(void*fn),$(int k), &data);
+      cpSpaceAddPostStepCallback($(cpSpace *space), $(void*fn),(void*) (uintptr_t) $(int k), &data);
     } |]
   pure ()

--- a/apecs/CHANGELOG.md
+++ b/apecs/CHANGELOG.md
@@ -1,3 +1,7 @@
+## [0.9.4]
+### Changed
+- (#86) Add support for GHC 9.2 and Template Haskell 2.18
+
 ## [0.9.3]
 ### Added
 - `newEntity_ = void . newEntity`

--- a/apecs/apecs.cabal
+++ b/apecs/apecs.cabal
@@ -1,5 +1,5 @@
 name:               apecs
-version:            0.9.3
+version:            0.9.4
 homepage:           https://github.com/jonascarpay/apecs#readme
 license:            BSD3
 license-file:       LICENSE
@@ -13,8 +13,8 @@ description:
   apecs is a fast, type-driven Entity-Component-System library for game programming.
 
 extra-source-files:
-  README.md
   CHANGELOG.md
+  README.md
 
 source-repository head
   type:     git

--- a/apecs/src/Apecs/THTuples.hs
+++ b/apecs/src/Apecs/THTuples.hs
@@ -91,13 +91,18 @@ tupleInstances n = do
 
       -- s, ety, w arguments
       sNs = [ mkName $ "s_" ++ show i | i <- [0..n-1]]
-      sPat = ConP tupleName (VarP <$> sNs)
       sEs = VarE <$> sNs
       etyN = mkName "ety"
       etyE = VarE etyN
       etyPat = VarP etyN
       wNs = [ mkName $ "w_" ++ show i | i <- [0..n-1]]
+#if MIN_VERSION_template_haskell(2,18,0)
+      sPat = ConP tupleName [] (VarP <$> sNs)
+      wPat = ConP tupleName [] (VarP <$> wNs)
+#else
+      sPat = ConP tupleName (VarP <$> sNs)
       wPat = ConP tupleName (VarP <$> wNs)
+#endif
       wEs = VarE <$> wNs
 
       getN     = mkName "ExplGet"
@@ -130,8 +135,13 @@ tupleInstances n = do
 
       explExistsAnd va vb =
         AppE (AppE (VarE '(>>=)) va)
+#if MIN_VERSION_template_haskell(2,18,0)
+          (LamCaseE [ Match (ConP 'False [] []) (NormalB$ AppE (VarE 'return) (ConE 'False)) []
+                    , Match (ConP 'True [] []) (NormalB vb) []
+#else
           (LamCaseE [ Match (ConP 'False []) (NormalB$ AppE (VarE 'return) (ConE 'False)) []
                     , Match (ConP 'True []) (NormalB vb) []
+#endif
                     ])
 
       explMembersFold va vb = AppE (VarE '(>>=)) va `AppE` AppE (VarE 'U.filterM) vb

--- a/cabal-ghc9.project
+++ b/cabal-ghc9.project
@@ -1,0 +1,13 @@
+-- The commented packages have dependencies that prevent them from building with
+-- GHC 9.2, and should be uncommented as soon as their dependencies have been
+-- updated. Once all of them are compatible again, we can remove this file.
+
+packages:
+  apecs
+  apecs-physics
+  -- apecs-gloss
+  -- apecs-stm
+  -- examples
+
+package apecs
+  tests: True

